gear/gear/auth.py

```diff
@@ -8,96 +8,101 @@
+        if not userdata:
+            raise web.HTTPUnauthorized()
+        return await fun(request, userdata, *args, **kwargs)
+    return wrapped
+
+
+def _web_unauthorized(request, redirect):
+    if not redirect:
+        raise web.HTTPUnauthorized()
+
+    login_url = deploy_config.external_url('auth', '/login')
+
+    # request.url is yarl.URL
+    request_url = request.url
+    x_forwarded_host = request.headers.get('X-Forwarded-Host')
+    if x_forwarded_host:
+        request_url = request_url.with_host(x_forwarded_host)
+    x_forwarded_proto = request.headers.get('X-Forwarded-Proto')
+    if x_forwarded_proto:
+        request_url = request_url.with_scheme(x_forwarded_proto)
+
+    raise web.HTTPFound(f'{login_url}?next={urllib.parse.quote(str(request_url))}')
+
+
+def web_authenticated_users_only(redirect=True):
     def wrap(fun):
         @wraps(fun)
         async def wrapped(request, *args, **kwargs):
-            def unauth():
-                if redirect:
-                    login_url = deploy_config.external_url('auth', '/login')
-
-                    # request.url is yarl.URL
-                    request_url = request.url
-                    x_forwarded_host = request.headers.get('X-Forwarded-Host')
-                    if x_forwarded_host:
-                        request_url = request_url.with_host(x_forwarded_host)
-                    x_forwarded_proto = request.headers.get('X-Forwarded-Proto')
-                    if x_forwarded_proto:
-                        request_url = request_url.with_scheme(x_forwarded_proto)
-
-                    raise web.HTTPFound(f'{login_url}?next={urllib.parse.quote(str(request_url))}')
-                raise web.HTTPUnauthorized()
-
-            if rest:
-                if 'Authorization' not in request.headers:
-                    unauth()
-                auth_header = request.headers['Authorization']
-                if not auth_header.startswith('Bearer '):
-                    unauth()
-                session_id = auth_header[7:]
-            else:
-                session = await aiohttp_session.get_session(request)
-                if 'session_id' not in session:
-                    unauth()
-                session_id = session['session_id']
-
-            headers = {'Authorization': f'Bearer {session_id}'}
-            try:
-                async with aiohttp.ClientSession(
-                        raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
-                    async with session.get(deploy_config.url('auth', '/api/v1alpha/userinfo'),
-                                           headers=headers) as resp:
-                        userdata = await resp.json()
-            except Exception:  # pylint: disable=broad-except
-                log.exception('getting userinfo')
-                unauth()
+            userdata = await userdata_from_web_request(request)
+            if not userdata:
+                _web_unauthorized(request, redirect)
```

Comment:
I think it would be clearer if we returned here, or just below, so that it's clear that no matter what_web_unauthorized does, we won't fall through